File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.py
import sys
from collections import deque
sys.setrecursionlimit(10**6)

input = sys.stdin.readline

# N : 도시의 개수, M : 도로의 개수, K : 거리 정보, X : 출발 도시 번호
N, M, K, X = map(int, input().strip().split())
graph = [[] for _ in range(N + 1)] # 인접 리스트
visited = [False] * (N + 1)

for _ in range(M):
    A, B = map(int, input().strip().split())
    graph[A].append(B) # 단방향 도로(무가중치)

result = [] # 최단거리가 K인 도시들(= 답)
def bfs(start):
    queue = deque([(start, 0)]) # (현재 도시, 현재까지 거리)
    
    visited[start] = True # 시작도시(X) 방문처리
    while queue:
        node, dist = queue.popleft()
        
        if dist == K:
            result.append(node)
            continue
        
        # 현재 도시에서 갈 수 있는 도시들 탐색
        for next_node in graph[node]:
            # 방문하지 않은 도시라면, 큐에 추가하고 방문처리
            if not visited[next_node]:
                visited[next_node] = True
                queue.append((next_node, dist + 1))
                
bfs(X)
if result:
    result.sort()
    sys.stdout.write('\n'.join(map(str, result)))
else:
    sys.stdout.write('-1')

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
const [N,M,K,X] = input[0].split(' ').map(Number);

// 큐 정의
class Queue {
    constructor() {
        this.items = new Map();
        this.headIndex = 0;
        this.tailIndex = 0;
    }
    
    enqueue(value) {
        this.items.set(this.tailIndex, value);
        this.tailIndex++;
    }
    
    dequeue() {
        if(this.isEmpty()) return null;
        
        const value = this.items.get(this.headIndex);
        this.items.delete(this.headIndex);
        this.headIndex++;
        
        return value;
    }
    
    isEmpty() {
        return (this.tailIndex === this.headIndex);        
    }
}

// 인접리스트
const graph = Array.from({ length: N+1 }, () => []);
for(let i=1; i<=M; i++) {
    const [from, to] = input[i].split(' ').map(Number);
    graph[from].push(to);
}

// BFS 수행
const dist = Array(N+1).fill(-1);
dist[X] = 0;
const q = new Queue();
q.enqueue(X);

while(q.isEmpty() !== true) {
    const current = q.dequeue();
    
    for(let next of graph[current]) {
        if(dist[next] === -1) {
            dist[next] = dist[current] + 1;
            q.enqueue(next);
        }
    }
}

// 결과 출력
const result = []; // K가 최단거리인 도시들
for(let i=1; i<=N; i++) {
    if(dist[i] === K) result.push(i);
}

console.log(result.length ? result.join('\n') : -1);